text/0000-reproducibility.md

```diff
@@ -0,0 +1,257 @@
+
+### Limited connectivity mode
+We may eventually want to offer a mode that users can opt into to use smaller platform packs, for situations where limited connectivity makes it hard to download the default packs. This should not be the default, and the UX should not *encourage* this mode, because it’s too tempting to avoid up-front downloads and then end up with less compression, more aggregate downloads and disk usage, and generally slower behavior. Also, users might be confused into thinking that the initial download is wasteful and spread misinformation about disk usage. Generally, the bet is that a multi-version git repository will lead to lower disk usage, so we should not encourage the use of more minimal packs. But as long as the UX is framed in terms of the use case of *limited connectivity* (for example, `install.sh --limited-connectivity`), this should help avoid misconceptions.
+
+Moreover, it should still be possible to upgrade a limited-connectivity install to a normal install, if the user finds themselves in better network conditions later on, without having to reinstall Notion from scratch.
+
+This particular feature is lower priority than the core functionality of this RFC and can be explored later.
+
+### Extensibility
+
+We will need a model for overriding the public git repositories, so that companies and CI systems can host their own Node distributions.
+
+# Critique
+[critique]: #critique
+
+It’s natural to question whether this implementation strategy is worth the effort: an alternative approach would be to allow projects to specify less precise version requirements (such as the ranges typically expressed in the `"engines"` field of `package.json`) and assume most differences will be benign. However, behavioral divergences between versions of Node do happen and are some of the trickiest bugs to nail down. Putting in extra work up front to ensure that these divergences *cannot happen, by construction* will eventually pay for itself when scaled across the Node ecosystem.
+
+Another reasonable criticism is that pinning the Node version for tools in the user toolchain means that users will not automatically benefit from performance and security improvements in Node. There are a couple of reasons this is outweighed by the benefits of pinning. First, as users upgrade the tool version itself, they will automatically get re-pinned to the newer version of Node specified by newer versions of the tool’s `package.json`. Second, we should at least allow users the option to override the tool’s specified Node version. But by letting the tool choose its platform version by default, user’s have a stronger guarantee that the tool will work and continue to work consistently based on how it was tested.
```

Comment:
This RFC is missing a discussion about installing user tools that do not have a pinned node version. Currently there are 0 tools that specify a `"toolchain"`, and even if Notion is wildly popular there will be a transition period where some tools have a pinned `node` version and some do not. Actually there may always be some tools that do not specify a `"toolchain"`, for whatever reason.

Notion should be able to handle this situation and preserve reproducibility. That could be using the version of node in the user's config, or if that is not specified, using the user's default version of node, and pinning that somehow. In these cases Notion could modify the installed tool's package.json to include a `"toolchain"` section if it is not already present.